core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -390,4 +386,43 @@ private void cleanExpiredSnapshots() {
 
     cleanupStrategy.cleanFiles(base, current);
   }
+
+  private void validateCleanupCanBeIncremental(TableMetadata current) {
```

Comment:
Don't we also need to check that the only table ref is `main`? This checks that there's one ref and only snapshots behind `currentSnapshot` are expired. But if `main` doesn't exist, then `allRemovedSnapshotsAreInMainAncestry` is going to fail with an NPE.

We also need to ensure that there were no non-main snapshots before expiration. Those need to be cleaned up through reachability analysis.

Reply:
I don't think we'd ever be in a state where only a non main ref is defined because at minimum the `main` branch would always be defined; even if it's empty. so the exactly one check is OK i think. I could make it more explicit though

>We also need to ensure that there were no non-main snapshots before expiration. Those need to be cleaned up through reachability analysis.

The ancestry check in the method in 407 would cover this because we'd check that we can't do incremental cleanup for any snapshots that are outside of main?

Reply:
Ah but even if a snapshot is not removed we still would need reachable analysis if there's a non-main snapshot since a staged but not removed snapshot could still reference a file that should still not be cleaned up

Reply:
I've amended this a bit so that we can distinguish where there only branch is main before expiration and after expiration we validate that there's no other refs other than main. That way so long as any tags on the ancestry of main age out we can incrementally cleanup, assuming the other conditions of no other non-main snapshots holds true. The non main snapshots are checked at the very end.

For the check where there are non main snapshots it handles the case where the current snapshot is null by simply checking if the whole `snapshots` list is not empty. If it's not empty then this means there are non-main snapshots.

Reply:
I think maybe we can do a little better. I think your argument for tags main's ancestors is a good one. We can do that for branches, too.

What about these rules to determine if incremental is safe?
1. No specified snapshots to remove
2. No snapshots to remove that are not ancestors of main
3. No snapshots after expiration that are not ancestors of main

I think the combination of 2 and 3 cover branch and tag cases. Tags function like branches with just one snapshot, so reasoning about branches will cover them. Any branch either has a snapshot outside of main or it doesn't. If it doesn't, then incremental cleanup will work. If there is a snapshot outside of main, then it is either going to be caught be 2 (if it ages off) or 3 (if it does not age off).

Reply:
After a lot of sketching out to prove to myself this works, I agree with the simplified algorithm. I went ahead and updated the implementation